lib/TTT/input.ex

```diff
@@ -3,76 +3,19 @@ defmodule Input do
     IO.stream(:stdio, :line)
   end
 
-  def user_move_to_internal_state(move) do
-    Map.get(
-      %{
-        1 => {0, 0},
-        2 => {0, 1},
-        3 => {0, 2},
-        4 => {1, 0},
-        5 => {1, 1},
-        6 => {1, 2},
-        7 => {2, 0},
-        8 => {2, 1},
-        9 => {2, 2}
-      },
-      String.to_integer(move),
-      :error
-    )
+  def clean(input) do
+    String.trim(input)
   end
 
-  def number?(move) do
-    try do
-      String.to_integer(move)
-      true
-    rescue
-      ArgumentError -> false
-    end
+  def analyze(move, state) do
+    with {:ok, :is_number} <- Validity.number?(move),
+         {:ok, :space_on_board} <- Validity.space_on_board?(move),
+         {:ok, :is_open} <- Validity.open?(move, state),
+         do: place_move(move, state)
```

Comment:
This is interesting.  Will this check each valid case, and then only progress to placing moves if they all pass, otherwise it's just a no-op?  Is that a correct read?

Reply:
Yes that's correct - if the patterns on each side of the arrow are a match it will proceed to the do block, if one of the lines doesn't match it instead returns the non-matching pattern, which in this case is the specific error message such as {:error, :not_a_number}